### ISSUE TITLE
(RoBERTa-base) Cannot reproduce CoLA and MRPC results (0.01x lower).

### ISSUE THREAD

**Issue by xuanqing94** [pytorch/fairseq#1445]
Hi I am trying to reproduce GLUE benchmarks with RoBERTa model. I have no problem with most of them except CoLA and MRPC datasets. I strictly followed the instructions in  [README.glue.md](https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.glue.md), and specifically for CoLA:
```
TOTAL_NUM_UPDATES=5336       
WARMUP_UPDATES=320      # 6 percent of the number of updates                                                                                                                                                                        
LR=1.0e-05                # Peak LR for polynomial LR scheduler.                                                                                                                                                                    
NUM_CLASSES=2                                                                                                                                                                                                                       
MAX_SENTENCES=4        # Batch size.                                                                                                                                                                                                
UPDATE_FREQ=1          # Update frequence                                                                                                                                                                                           
ROBERTA_PATH=./checkpoints/roberta.base/model.pt                                                                                                                                                                                    
SAVE_DIR=./checkpoints/cola                                                                                                                                                                                                                                                                                                                                                                                                                                    
CUDA_VISIBLE_DEVICES=2,3,6,7 python ../fairseq/train.py CoLA-bin/ \                                                                                                                                                                     
--restore-file $ROBERTA_PATH \                                                                                                                                                                                                      
--max-positions 512 \                                                                                                                                                                                                               
--max-sentences $MAX_SENTENCES \                                                                                                                                                                                                    
--max-tokens 4400 \                                                                                                                                                                                                                 
--task sentence_prediction \                                                                                                                                                                                                        
--reset-optimizer --reset-dataloader --reset-meters \                                                                                                                                                                               
--required-batch-size-multiple 1 \                                                                                                                                                                                                  
--init-token 0 --separator-token 2 \                                                                                                                                                                                                
--arch roberta_base \                                                                                                                                                                                                               
--criterion sentence_prediction \                                                                                                                                                                                                   
--num-classes $NUM_CLASSES \                                                                                                                                                                                                        
--dropout 0.1 --attention-dropout 0.1 \                                                                                                                                                                                             
--weight-decay 0.1 --optimizer adam --adam-betas "(0.9, 0.98)" --adam-eps 1e-06 \                                                                                                                                                   
--clip-norm 0.0 \                                                                                                                                                                                                                   
--lr-scheduler polynomial_decay --lr $LR --total-num-update $TOTAL_NUM_UPDATES --warmup-
updates $WARMUP_UPDATES \                                                                                                                   
--max-epoch 10 \                                                                                                                                                                                                                    
--update-freq $UPDATE_FREQ \                                                                                                                                                                                                        
--find-unused-parameters \                                                                                                                                                                                                          
--best-checkpoint-metric accuracy --maximize-best-checkpoint-metric \                                                                                                                                                               
--save-dir $SAVE_DIR \                                                                                                                                                                                                              
--fp16 --fp16-init-scale 4 --threshold-loss-scale 1 --fp16-scale-window 128 \ 

```
I made two independent runs, the accuracy was 0.84, and matthews_corrcoef was 0.619. I calculate mattews corrcoef with sklearn package. The reported value at https://github.com/pytorch/fairseq/tree/master/examples/roberta is 0.636 mattews corrcoef.

For MRPC, I got 0.889 accuracy and 0.92 F1. While the reported value is 0.90 accuracy. Still a few digits bigger than my reproduction.

Is the performance gap due to random effect or incorrect hyperparameter settings?

**Comment by ypengc7512** [comment]
Hello, I also encountered the same problem. I made six independent runs and the matthews_corrcoef was 0.615 for roberta-base on the CoLA task. How did you solve that?

**Comment by xuanqing94** [comment]
@ypengc7512 I changed the random seed (but I forgot the exact number that works).

**Comment by CSerxy** [comment]
Hi, I wonder would you mind sharing your for inference. My code can achieve the reported accuracy but not for matthews_corrcoef. Accuracy for RoBERTa-base is 0.835, and matthews_corrcoef is 0.596. Here is my code:

`from fairseq.models.roberta import RobertaModel
from sklearn.metrics import matthews_corrcoef

roberta = RobertaModel.from_pretrained(
    '~/roberta.base/cola-finetune',
    checkpoint_file='checkpoint_best.pt',
    data_name_or_path='CoLA-bin'
)

label_fn = lambda label: roberta.task.label_dictionary.string(
    [label + roberta.task.label_dictionary.nspecial]
)
ncorrect, nsamples = 0, 0
roberta.cuda()
roberta.eval()
y_true, y_pred = [], []
with open('glue_data/CoLA/dev.tsv') as fin:
    for index, line in enumerate(fin):
        tokens = line.strip().split('\t')
        sent, target = tokens[-1], tokens[1]
        tokens = roberta.encode(sent)
        prediction = roberta.predict('sentence_classification_head', tokens).argmax().item()
        prediction_label = label_fn(prediction)
        ncorrect += int(prediction_label == target)
        y_true.append(int(target))
        y_pred.append(int(prediction_label))
        nsamples += 1
print('matthews_corrcoef: ' + str(matthews_corrcoef(y_true, y_pred)))
print('ncorrect / nsamples = ' + str(ncorrect) + ' / ' + str(nsamples))
print('| Accuracy: ', float(ncorrect)/float(nsamples))
`

**Comment by xuanqing94** [comment]
@CSerxy 
```python
import numpy as np
from sklearn.metrics import matthews_corrcoef
from fairseq.models.roberta import RobertaModel

roberta = RobertaModel.from_pretrained(
    'checkpoints/flow/cola',
    checkpoint_file='checkpoint_best.pt',
    data_name_or_path='CoLA-bin'
)

label_fn = lambda label: roberta.task.label_dictionary.string(
    [label + roberta.task.label_dictionary.nspecial]
)
ncorrect, nsamples = 0, 0
predictions = []
ground_truth = []
roberta.cuda()
roberta.eval()
with open('glue_data/CoLA/dev.tsv', encoding='utf-8') as fin:
    fin.readline()
    for index, line in enumerate(fin):
        tokens = line.strip().split('\t')
        sent, target = tokens[3], tokens[1]
        tokens = roberta.encode(sent)
        prediction = roberta.predict('sentence_classification_head', tokens).argmax().item()
        prediction_label = label_fn(prediction)
        ncorrect += int(prediction_label == target)
        prediction_label = int(prediction_label)
        target = int(target)
        predictions.append(prediction_label)
        ground_truth.append(target)
        nsamples += 1

print('| Accuracy: ', float(ncorrect)/float(nsamples))
MCC = matthews_corrcoef(ground_truth, predictions)
print('| MCC: ', MCC)
```
